File: filter/source/svg/gentoken.pl
# from oox/source/token - should really put this into solenv

$ARGV0 = shift @ARGV;
$ARGV1 = shift @ARGV;
$ARGV2 = shift @ARGV;

open ( TOKENS, $ARGV0 ) || die "can't open token file: $!";
my %tokens;

while ( defined ($line = <TOKENS>) )
{
    if( !($line =~ /^#/) )
    {
        chomp($line);
        @token = split(/\s+/,$line);
        if ( not defined ($token[1]) )
        {
            $token[1] = "XML_".$token[0];
            $token[1] =~ tr/\-\.\:/___/;
            $token[1] =~ s/\+/PLUS/g;
            $token[1] =~ s/\-/MINUS/g;
        }

        $tokens{$token[0]} = uc($token[1]);
    }
}
close ( TOKENS );

open ( HXX, ">$ARGV1" ) || die "can't open tokens.hxx file: $!";
open ( GPERF, ">$ARGV2" ) || die "can't open tokens.gperf file: $!";

print ( GPERF "%language=C++\n" );
print ( GPERF "%global-table\n" );
print ( GPERF "%null-strings\n" );
print ( GPERF "%struct-type\n" );
print ( GPERF "struct xmltoken\n" );
print ( GPERF "{\n" );
print ( GPERF "  const sal_Char *name; sal_Int32 nToken; \n" );
print ( GPERF "};\n" );
print ( GPERF "%%\n" );

print ( HXX "#ifndef INCLUDED_AUTOGEN_TOKEN_HXX\n" );
print ( HXX "#define INCLUDED_AUTOGEN_TOKEN_HXX\n\n" );
print ( HXX "#include <sal/types.h>\n\n" );

$i = 0;
foreach( sort(keys(%tokens)) )
{
    print( HXX "const sal_Int32 $tokens{$_} = $i;\n" );
    print( GPERF "$_,$tokens{$_}\n" );
    $i = $i + 1;
}
print ( GPERF "%%\n" );
print ( HXX "const sal_Int32 XML_TOKEN_COUNT = $i;\n" );
print ( HXX "const sal_Int32 XML_TOKEN_INVALID = -1;\n\n" );
print ( HXX "#endif\n" );
close ( HXX );
close ( GPERF );

File: filter/source/svg/gentoken.py
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import sys

tokenfile_name = sys.argv[1]
hxx_name = sys.argv[2]
gperf_name = sys.argv[3]

gperf_header = r"""%language=C++
%global-table
%null-strings
%struct-type
struct xmltoken
{
  const char *name; sal_Int32 nToken;
}
%%
"""

tokens = {}

with open(tokenfile_name) as tokenfile:
   for line in tokenfile:
       line = line.strip()
       if line:
           arr = line.split()
           if len(arr) < 2:
               t = "XML_" + arr[0]
               t = t.replace('-', '_').replace('.', '_').replace(':', '_')
               t = t.replace('+', 'PLUS')
               arr.append(t)
           tokens[arr[0]] = arr[1].upper()

hxx = open(hxx_name, 'w')
gperf = open(gperf_name, 'w', newline='\n')

gperf.write(gperf_header)

hxx.write("#ifndef INCLUDED_AUTOGEN_TOKEN_HXX\n")
hxx.write("#define INCLUDED_AUTOGEN_TOKEN_HXX\n\n")
hxx.write("#include <sal/types.h>\n\n" )

i = 0
for token in sorted(tokens.keys()):
    i += 1
    hxx.write("const sal_Int32 {} = {};\n".format(tokens[token], i))
    gperf.write("{},{}\n".format(token, tokens[token]))

gperf.write("%%\n")
hxx.write("const sal_Int32 XML_TOKEN_COUNT = {};\n".format(i))
hxx.write("const sal_Int32 XML_TOKEN_INVALID = -1;\n\n")
hxx.write("#endif\n")

hxx.close()
gperf.close()
